gtn/functions/compose.cpp

```diff
@@ -450,3 +451,665 @@ Graph compose(
+
+    // Calculate number of threads we have to spawn
+    std::vector<std::pair<int.int>> toExploreNumEdge(toExploreNodePair.size());
+    std::vector<int> edgeMulOffset(toExploreNodePair.size());
+
+    // No dependence between iterations
+    for (int i = 0; i < toExploreNodePair.size(); ++i) {
+      int node = toExploreNodePair[i].first;
+      const int numEdgesFirst =
+          graphDP1.inArcOffset[node + 1] - graphDP1.inArcOffset[node];
+
+      node = toExploreNodePair[i].second;
+      const int numEdgesSecond =
+          graphDP2.inArcOffset[node + 1] - graphDP2.inArcOffset[node];
+
+      toExploreNumEdge[i] = std::make_pair(numEdgesFirst, numEdgesSecond);
+      edgeMulOffset[i] = numEdgesFirst * numEdgesSecond;
+    }
+
+    prefixSumScan(edgeMullOffset);
+    assert(!edgeMulOffset.empty());
+    const int totalEdges = edgeMullOffset[edgeMullOffset.size() - 1];
+
+    // No dependence between iterations. tid is thread-id
+    // Only do non epsilon case for this kernel
+    for (int tid = 0; tid < totalEdges; ++tid) {
+      // Node pair
+      std::pair<int, int> nodePair;
+      std::pair<int, int> edgePair;
+
+      // Map tid to corresponding node and edge pair
+      // Search to find which node pair this tid will fall into
+      // Linear search for now (edgeMullOffset is sorted by definition)
+      std::vector<int, int> nodePair;
+      std::vector<int, int> edgePair;
+      std::tie(nodePair, edgePair) =
+          computeNodeAndEdgePair(tid, edgeMullOffset, toExploreNodePair);
+
+      // Does this node pair match?
+      if (graphDP1.olabels[edgePair.first] ==
+          graphDP2.ilabels[edgePair.second]) {
+        const int idx = TwoDToOneDIndex(
+            graphDP1.srcNodes[edgePair.first],
+            graphDP2.srcNodes[edgePair.second],
+            numNodesFirst);
+        if (!reachable[idx]) {
+          toExplore[idx] = true;
+        }
+        reachable[idx] = true;
+
+        if (graphDP1.olabels[edgePair.first] == epsilon) {
+          epsilonMatched[TwoDToOneDIndex(
+              nodePair.first, nodePair.second, numNodesFirst)] = true;
+        }
+      }
+    }
+
+    // No dependence between iterations. tid is thread-id
+    // Do epsilon match case in this kernel launch
+    for (int tid = 0; tid < totalEdges; ++tid) {
+      // Map tid to corresponding node and edge pair
+      // Search to find which node pair this tid will fall into
+      // Linear search for now (edgeMullOffset is sorted by definition)
+      std::vector<int, int> nodePair;
+      std::vector<int, int> edgePair;
+      std::tie(nodePair, edgePair) =
+          computeNodeAndEdgePair(tid, edgeMullOffset, toExploreNodePair);
+
+      const bool matched = epsilonMatched[TwoDToOneDIndex(
+          nodePair.first, nodePair.second, numNodesFirst)];
+      // Note that when this loop body is threaded - multiple threads will
+      // write to the same location. They will write the same value so
```

Comment:
`std::vector<bool>` is, I believe, implemented as a bitmap, so I doubt it's that simple.  I'd use std::vector<char> instead here.